nym-connect/desktop/src/context/main.tsx

```diff
@@ -162,39 +167,60 @@ export const ClientContextProvider: FCWithChildren = ({ children }) => {
   const shouldUseUserGateway = !!userDefinedGateway.gateway && userDefinedGateway.isActive;
   const shouldUseUserSP = !!userDefinedSPAddress.address && userDefinedSPAddress.isActive;
 
-  const setServiceProvider = async (newServiceProvider: ServiceProvider) => {
-    await invoke('set_gateway', {
-      gateway: shouldUseUserGateway ? userDefinedGateway.gateway : newServiceProvider.gateway,
-    });
+  const applyServiceProvider = async (newServiceProvider: ServiceProvider) => {
     await invoke('set_service_provider', {
       serviceProvider: shouldUseUserSP ? userDefinedSPAddress.address : newServiceProvider.address,
     });
   };
 
+  const applyGateway = async (newGateway: Gateway) => {
+    await invoke('set_gateway', {
+      gateway: shouldUseUserGateway ? userDefinedGateway.gateway : newGateway.identity,
+    });
+  };
+
   const getRandomSPFromList = (services: ServiceProvider[]) => {
     const randomSelection = services[Math.floor(Math.random() * services.length)];
     return randomSelection;
   };
 
+  const getRandomGatewayFromList = (theGateways: Gateway[]) => {
+    const randomSelection = theGateways[Math.floor(Math.random() * theGateways.length)];
+    return randomSelection;
+  };
+
   const buildServiceProvider = async (serviceProvider: ServiceProvider) => {
     const sp = { ...serviceProvider };
-
-    if (shouldUseUserGateway) sp.gateway = userDefinedGateway.gateway as string;
     if (shouldUseUserSP) sp.address = userDefinedSPAddress.address as string;
-
     return sp;
   };
 
-  const setSerivceProvider = async () => {
+  const buildGateway = async (gateway: Gateway) => {
+    const gw = { ...gateway };
+    if (shouldUseUserGateway) gw.identity = userDefinedGateway.gateway as string;
+    return gw;
+  };
+
+  const setServiceProvider = async () => {
     if (serviceProviders) {
       const randomServiceProvider = getRandomSPFromList(serviceProviders);
       const withUserDefinitions = await buildServiceProvider(randomServiceProvider);
-      await setServiceProvider(withUserDefinitions);
+      await applyServiceProvider(withUserDefinitions);
       setSelectedProvider(withUserDefinitions);
     }
     return undefined;
   };
 
+  const setGateway = async () => {
+    if (gateways) {
+      const randomGateway = getRandomGatewayFromList(gateways);
+      const withUserDefinitionsForGateway = await buildGateway(randomGateway);
+      await applyGateway(withUserDefinitionsForGateway);
+      // Do we need a setSelectedGateway?
```

Comment:
"Do we need a setSelectedGateway?" If Im not wrong I think the answer is no
@fmtabbara can you confirm?

Reply:
Added in setSelectedGateway, it was needed for the updated tooltip